### DISCUSSION TITLE
Nilan Compact P Air 9 support

### DISCUSSION THREAD

**Discussion by SalifTraoure** [veista/nilan#115] [General]
Hi,

I have a Compact P Air 9 unit with CTS602 control. 
Any plans to support this unit any time soon ? 

I have the communication set up to HA and read the following data. 
HW ID: 16
Application id: -32756

I was emailed the attached mapping by Nilan to be the correct one ( althought the document  states : 32780=combi). 
I haven't challenged them on it being wrong. Most registers seems to be correct. 

( Although date and timers decoding is a challenge for a newbie like me ). That's why I was hoping my unit would be supported.

[CombiController_RemoteModbus v4.pdf](https://github.com/veista/nilan/files/13908232/CombiController_RemoteModbus.v4.pdf)

**Comment by veista** [comment]
Follow the instructions in the read me file is the plan.

**Reply by SalifTraoure** [reply]
Enabling debugging mode for HACS and reproducing the setup doesn't seem to produce anything in the log. 
I get the feedback "Unsupported Device". See attached photos for instructions given in readme. 
Sorry for not noticing those.. 

![nameplate](https://github.com/veista/nilan/assets/122399688/5dfba8f3-8b3b-4e27-97c8-2e25ca8fb0ac)
![device-info](https://github.com/veista/nilan/assets/122399688/6637f830-948a-4313-8ec3-cab1b2764397)
[home-assistant_hacs_2024-01-11T21-53-43.413Z.log](https://github.com/veista/nilan/files/13908818/home-assistant_hacs_2024-01-11T21-53-43.413Z.log)

**Reply by veista** [reply]
I need a picture of the HMI showing the device name. You should enable debug logging for nilan integration, not HACS. Read previous issues and conversations.

**Reply by veista** [reply]
That is the right page on HMI, Your model seems to be different from all the others in that regard, Enable the logging and let's see what that brings up.

**Reply by veista** [reply]
PS. If it was not obvious to you, do not have your own modbus instance running while trying to communicate via this integration.

**Comment by veista** [comment]
I added support for your device in the latest beta. Be sure to have betas enabled in HACS if you are using HACS.

Please do the following:
1. Report your modbus version shown by the integration
2. Report any unresponsive/non-working entities
3. Cross-check all functionality with the HMI panel

**Reply by SalifTraoure** [reply]
Hi,

Thanks for making the effort to add the support. I really appreciate it. 
( I didn't come back to you before as I couldn't find a specific debugging option for the Nilan integration) 

Here are the answers to your question: 
1. Modbus version 2, Firmware 112
2.  Seems to be working. I see a discprenacy between the HMI value of outdoor temperature. I belive the reading is correct, but my display is configured to be using the outdoor temperature from the Air unit which is outdoors (T20?) and more accurate. In my configuration the travel distance of duct is quite long so there was 4.6degC outdoor is warmed up to 6.4degC as fresh air intake.  No issue, just mention it.

3. What is there seems ok as far as I could see.  The "display LED 1/2" I have no clue what is. Apart from that 32 entities discovered ( attached photo) 
   Values/commands from display homepage that I could not found to be available are.
   - Return Water temperature (central heating)
   - Supply water minimum temperature (central heating setpoint) 
   - HotWater temperature top
   - HotWater temperature bottom
   - HotWater temperature setpoint
   I was hoping there was command features, but it doesn't seem to be the case. 

![nilan_screenshot](https://github.com/veista/nilan/assets/122399688/4aea59b8-dc24-42d3-9237-d35bca5fb8db)
![screenshot diagnostic](https://github.com/veista/nilan/assets/122399688/f71c3186-d8d5-4c74-abbe-729043dee75b)
![image (1)](https://github.com/veista/nilan/assets/122399688/053bd421-136b-4879-83cf-5c5646a562e1)

Note: My initial motivation to get the modbus up and running was to log values as part of troubleshooting a problem I have with shutdowns. I therefore need to track other variables like the outdoor AIR unit compressors state, FanCoil output (voltage) and pressurepipe temp (T35), in order to provide data for technicians to understand what is going on.. 
I guess I may need to inspect and modify your code ( which is probably not realistic) - or revert to my basic reading of the values I need to solve my problem. 

Any plans to add commands by the way?

**Reply by veista** [reply]
Thanks for the reply. Your model seems to be odd. Having modbus version 2 with a touch panel -> doesn't follow regular Nilan logic.

I will try to fix these issues and make a new beta for you to test.

**Reply by veista** [reply]
So you dont have any controls? Water heaters or climate? Try to refresh the page.

**Reply by SalifTraoure** [reply]
No, no other item types than sensors.
I tried refresh / restart / and clearing cache. 

I was struggling to use the build in HA 'climate' type as Nilan don't support FC16 writing to multiple registers which is the default code used even if it is only writing one register. I had to use a work around with an automation to write a single register. 
I guess you have a workaround in your code to bypass this limitation?

**Reply by veista** [reply]
Can you check the logs for any errors?

**Reply by SalifTraoure** [reply]
Some errors have been reoccurring since it was installed yesterday.
It doesn't seem that the installation process itself encounter any errors that are logged. 


[home-assistant_nilan_2024-02-05T18-34-25.529Z_install_debug_active.log](https://github.com/veista/nilan/files/14170154/home-assistant_nilan_2024-02-05T18-34-25.529Z_install_debug_active.log)
[home-assistant_nilan_2024-02-05T18-30-35.481Z_since yesterday.log](https://github.com/veista/nilan/files/14170155/home-assistant_nilan_2024-02-05T18-30-35.481Z_since.yesterday.log)

**Reply by veista** [reply]
Thank you! It is going to be a lot of work getting your device properly supported. I need to add some new logic since this combi controller uses somewhat different logic. This is why you don't have any controls.

**Reply by veista** [reply]
In the mean time, can you list all **HPS registers**, from the document you provided, you think should be implemented?

**Reply by veista** [reply]
One important question, can you read the T35 sensor with your implementation?

**Reply by veista** [reply]
Give the latest release a try. According to the document T35 requires bus version 3, you have 2. If the document is wrong, please do tell.

**Reply by SalifTraoure** [reply]
T35 works fine when I read it manually as a modbus sensor with my own YAML.

I did my best to hightlight the registers I believe is relevant in the attached pdf. 
Green: 1st priority
- Basic data
+ needed for my troubleshooting / diagnostic

Yellow: 2nd priority
- Convenient, but strictly not necessary
Alarms, Timers

[CombiController_RemoteModbus.v4.pdf](https://github.com/veista/nilan/files/14172683/CombiController_RemoteModbus.v4.pdf)

I am not 100% sure it is the correct document although this was the document I received from Nilan support when I struggled to find the right one online.  My device reports : Application ID 32756, whilst the document states application id:32780=Combi. 
I will ask Nilan to be 100% sure, but in my experience they take weeks to answer. 
From reading back values - it seems to be correct though... 

The only thing I don't understand is the "setpoint" for central heating - water circuits. In support they reverted to me and said I should use holding register 1802 if the mode was compressor minimum. But reading back 1802 it doesn't match with HMI, so I'm struggling with this. 

![flow2](https://github.com/veista/nilan/assets/122399688/b93396a5-a80b-4ab8-83ad-532cdeea7acd)
![flow1](https://github.com/veista/nilan/assets/122399688/df210f6b-e52c-43d7-9357-c85939871808)

Regarding the updated version - Still no climates. Still 32 entities which are sensors. Same as before. 
Logs look the same as well.

**Reply by veista** [reply]
Are you 100% sure you tried 1.2.13B2 version?

**Reply by veista** [reply]
Those alarm code implemtations might be really tricky. Otherwise should be straight forward once the basic functionality works.

**Reply by veista** [reply]
Also you might have to "re-install" after update.

**Reply by veista** [reply]
Ok. I see the issue. Your hw type is 44, not 16. This complicates things immensly. I will try to figure something out...

You can test this by using version 1.2.12. If it "works" then that is the issue.

**Comment by SalifTraoure** [comment]
Register 5 reports 16 which should be HwID.  Where did you detect HWid 44? 
I can ask Nilan for clarity if in doubt

![image](https://github.com/veista/nilan/assets/122399688/08c8e9b0-1366-46e8-b8de-6474140a648a)

I tested version 1.2.12 ( reinstall custom Nilan integration and restart HA --> add device). Same result. Same 32 entities. Only sensors.

**Reply by veista** [reply]
HWId is the CTS602 board revision. I have 16 also. Machine type is in register HR1000.

**Reply by SalifTraoure** [reply]
Understood and yes, I can see that HR1000 = 44. 
By the way, I saw that I wrote wrong before. My device ( and probably other Nilan devices) does not support Modbus FC6 - write single register which I understand is what is used by the built in climate function.

**Reply by veista** [reply]
Just released latest candidate. Hope this one works.

**Reply by SalifTraoure** [reply]
I believe I used the latest one - there is no information displaying specifically which version - only that your Nilan integration was updated 4 minutes ago. 

See attached log. 
[home-assistant_nilan_2024-02-06T17-06-42.086Z.log](https://github.com/veista/nilan/files/14183893/home-assistant_nilan_2024-02-06T17-06-42.086Z.log)

By the way - I also finally identified the correct HR to set the setpoint for the central heating water system (as set through the display) 
It is HR6562 HPS_Heating.SetPointMin.  ( with IR1900 as the measurement feedback - Supply temperature)

**Reply by veista** [reply]
Ok. I will add some debug logging. Be sure to enable it in config yaml this time :)

**Reply by veista** [reply]
Latest version is 1.2.13B3

**Reply by veista** [reply]
Are you using HACS?

**Reply by veista** [reply]
https://github.com/veista/nilan/releases

**Reply by SalifTraoure** [reply]
Yes, I am using HACS. 
I will refresh my memory on how to actually do debugging. ( like you say - enable it config.yaml)

**Reply by SalifTraoure** [reply]
I might have been fooled by the HACS / Nilan stating *updated 4minutes ago* when I reinstalled.
I will try the manual installation of the latest pre-release in an hour or two. Need to get some kids to bed. 

I will add the following to configuration.yaml in addition.
logger:
  default: critical
  logs:
    # log level for a custom integration
    custom_components.nilan: debug

**Reply by veista** [reply]
Ok, so now I added the debugging strings to release 1.2.13B4

**Reply by SalifTraoure** [reply]
Good news. No errors. 104 entities appeared. 
I think I spotted one value that seems to read the wrong register. I will go through and update later tonight.

[home-assistant_nilan_2024-02-06T18-53-28.715Z.log](https://github.com/veista/nilan/files/14184937/home-assistant_nilan_2024-02-06T18-53-28.715Z.log)

**Reply by SalifTraoure** [reply]
I did basically a full check of all values listed in the home view. Some scaling findings - registers seems to be correct.  See attached png.
I did some few corrections to some descriptions ( could also be mistake in the local HMI :-)  - so don't bother to update descriptions if it matches document)

![Nilan - value check](https://github.com/veista/nilan/assets/122399688/36c70d56-68bf-45fd-a1f1-3ad6d932477a)

As for the other settings amongst the 104 entities that appeared - I didn't check them. 
But the ones I checked seemed ok. 
Just one unit mistake on ventilation time reported as seconds should be minute - But doesn't really matter. 
![ventilation-unit](https://github.com/veista/nilan/assets/122399688/c85ba6cc-bc37-48ce-99bd-9aef9acee43c)

EDIT - deleted humidty part - all ok. Understood the setpoint being the threshold for low humidity setting.

**Comment by SalifTraoure** [comment]
1 more observation

Is it a quick fix to add the FC voltage reading? , IR1925, ( Scale 0.1 - checked)  
![image](https://github.com/veista/nilan/assets/122399688/67b25797-0a91-442a-acc4-b9fcccdcc5ee)

**Reply by SalifTraoure** [reply]
Edit:  alam - misunderstanding..

**Reply by veista** [reply]
I think you misunderstand. That does not show hps alarms, only the indoor unit alarms.

**Reply by SalifTraoure** [reply]
Understood. Thanks for clarifying. I also edited / deleted the comment about fan controls in case you read it. I realized it was incorporated in the climate module.

**Comment by veista** [comment]
I made a new release for you to try on sunday. Did it fix your issues?

**Reply by SalifTraoure** [reply]
Hello again. All good with the latest release! I consider this to be a fully functional integration! 3 more entities was added. GEO alarm and Voltage as per my request. Thanks a lot. 
I guess the GEO/AIR alarm reset command is a separate HR which you could consider to add in the future for consistency. 
![image](https://github.com/veista/nilan/assets/122399688/0f7e393a-6aec-478a-aaa4-c19ceb4547a1)

The only other observation which I don't necessarily think is wrong is the efficiency going beyond 100%. I guess this is a calculation (maybe affected by the amount of duct going inside my house for supply air affecting the calculation...? ) 
Anyway - it's not something I would care to monitor. Just to mention it..  

Value check - both BETAs. 
![Nilan - value check-updated](https://github.com/veista/nilan/assets/122399688/0988ad27-4d6c-4966-aa2c-366606ecff9f)

**Reply by veista** [reply]
Yes efficiency is compared to electricity 1000W electricity gets you 2000 W of heat -> efficiency 200% it is silly, I know. I can try adding the alarm reset, all alarms seem simple.

**Reply by veista** [reply]
Once you get the alarms active, read the hps alarm id registers contents.

**Comment by SalifTraoure** [comment]
Got one alarm - See readings as follows:

![image](https://github.com/veista/nilan/assets/122399688/91bbf0dd-86bb-4fda-8c82-fb397f39104f)

Alarms follow the Software guide (page 45 - 47) 

[Software - NYESTE - 2021-06 S24 Compact P AIR DK.pdf](https://github.com/veista/nilan/files/14327955/Software.-.NYESTE.-.2021-06.S24.Compact.P.AIR.DK.pdf)

Let me know if I can contribute in any way by putting the alarm id's / descriptions / guidance into any format. I guess I would look into adding the guidance ( 3rd colum ) - into the description - to have the guidance on what to do than rather just the name of the alarm.

**Reply by veista** [reply]
Hi,

Great, I will try adding the alarm support when I have time.

**Reply by SalifTraoure** [reply]
Sounds good. There is no urgency for it. I'm very grateful for everything you've done and I'll be sending some coffee your way :-) 


I noticed the allowed range in local HMI is 5-50 for all of these
![image](https://github.com/veista/nilan/assets/122399688/ce73ddd1-eb86-4605-a4df-02ed07fd5a9b)

Whilst there are a more narrow range configured in some input validation code. 
I only detected this because I was playing with the settings as my unit seems to ignore these limits.  The ranges defined as input validation restrictions seems rather sensible. It was just to let you know. 

![image](https://github.com/veista/nilan/assets/122399688/f85cc636-e367-492e-a327-e2690980378a)

**Reply by veista** [reply]
Any support is welcome, thank you!

Thanks for noticing this.

**Reply by veista** [reply]
https://github.com/veista/nilan/releases/tag/1.2.13B6

**Reply by veista** [reply]
Can you test if the reset logic works with hps errors? I won't merge to master until you say it works, thanks.

**Reply by SalifTraoure** [reply]
Thanks. I was absent on winter break last week. Installed today. Will try to provoke an alarm. I didn't see any change in number of entities (106) so I guess you created the reset logic to reset both GEO/AIR and HVAC alarms through a common function ? ( although I believe it was two different registers) 

![image](https://github.com/veista/nilan/assets/122399688/c721973b-983d-4817-8654-341951946026)

**Reply by veista** [reply]
Should come up as its own entity. Reset AIR/GEO Alarm

**Reply by veista** [reply]
Can you check the logs for any errors?

**Comment by SalifTraoure** [comment]
I deleted the configuration. Updated all the files from B6. 
Reinstalled the integration. 
I thought this was sufficient. But it seemed a reset did the job. Now showing 107 entities and the AIR/GEO reset. 

The debug revelealed another error 
2024-02-24 16:40:07.138 ERROR (MainThread) [custom_components.nilan.device] Could not read get_external_heating_offset
[home-assistant_nilan_2024-03-05T15-36-15.432Z.log](https://github.com/veista/nilan/files/14497834/home-assistant_nilan_2024-03-05T15-36-15.432Z.log)

Am I missing another offset? 

![image](https://github.com/veista/nilan/assets/122399688/aa1af938-023d-4ce0-a566-d90bd00b7dab)

**Reply by veista** [reply]
Hi,

Great.

That means that your device does not physically have that register or just some error occured reading it. I will disable the entity for your device.

**Reply by veista** [reply]
How many occurances does the error have?

**Reply by veista** [reply]
BTW the entity that had that single error was "Room Temperature Neutral Zone", I think that is just a regular read error.

**Comment by SalifTraoure** [comment]
Hi. Finally managed to provoke an alarm. The reset seems to be working! Well done!  
I do sometimes have to try several times and give it like 20 seconds to react. If I tried to quick I would get error message like you can see in screenshot (at the bottom). I guess it is a combination of Modbus polling delay effect that is natural. I also have this somewhat tricky behaviour when resetting the alarms from the local display by having to try multiple times, It may be an issue/bug with my firmware in the way it's processing alarms. 

Is the reset code checking whether the alarm is 'inactive - and 'ready to be reset' - or if it is still active and cannot be reset? 
I noticed this from the local HMI which reports them as 'inactive'  (I guess this is the IR1602)

In an attempt to see the timing for when the alarm came in I was also looking in the logs to see if the Alarm ID was recorded anywhere else than in the Reset command window - like in the logbook. With 2 alarms incoming I was curious to see which one came first and if the second is a domino effect. I could not see an entry in the logbook but could see the rising edge in trend for active alarms. Also for troubleshooting and seeing patterns I get no record of which alarms that have been seen unless I keep a manual record on the side once I reset them. (It would be good to be able to document and show alarm patterns to Nilan for a support purposes) 

![unnamed](https://github.com/veista/nilan/assets/122399688/e1ad17d7-fa38-4d47-b5f4-5b5a581704f8)
![0](https://github.com/veista/nilan/assets/122399688/05a42e11-5363-4a21-b559-b79b9db643f0)
![0](https://github.com/veista/nilan/assets/122399688/ec80ce7e-5c5a-47df-9351-eb33f524752e)
